internal/graph/graph.go

```diff
@@ -137,16 +137,44 @@ func (g *Graph) InsertConstructor(ctor interface{}) error {
 	return nil
 }
 
-// ValidateReturnTypes validates if ctor's return type is already insterted in the graph
-func (g *Graph) ValidateReturnTypes(ctype reflect.Type) error {
-	objMap := make(map[reflect.Type]bool, ctype.NumOut())
+// ValidateInvokeReturnTypes validates Invoke return types and returns an error
+// if the grah node of return type is resolved and cached.
+func (g *Graph) ValidateInvokeReturnTypes(ctype reflect.Type) error {
+	if err := g.checkDuplicateReturns(ctype); err != nil {
+		return err
+	}
 	for i := 0; i < ctype.NumOut(); i++ {
 		objType := ctype.Out(i)
 		if _, ok := g.nodes[objType]; ok {
-			return errors.Wrapf(errRetNode, "ctor: %v, object type: %v", ctype, ctype.Out(i))
+			if obj, ok := g.nodes[objType].(*objNode); ok && obj.cached {
```

Comment:
This does the same lookup in g.nodes, as in previous if above, just use a variable there.